pkg/walker/walker.go

```diff
@@ -27,60 +28,54 @@ type WalkJob struct {
 
 func Run(jobs []WalkJob, par int, bufferSize int, verbose int) []string {
 	defer utils.TimeTrack(time.Now(), "walker run", verbose)
-	if len(jobs) == 0 {
+	initialJobCount := len(jobs)
+	if initialJobCount == 0 {
 		return []string{}
 	}
-	spawn := make(chan WalkJob, bufferSize)
-	start := make(chan bool, bufferSize)
-	end := make(chan []string)
 
-	for _, j := range jobs {
-		spawn <- j
-	}
+	results := make(chan []string, bufferSize)
+	jobQueue := make(chan WalkJob, bufferSize)
+	spawn := make(chan WalkJob, bufferSize)
 
-	for w := 1; w <= par; w++ {
-		go worker(w, spawn, start, end, verbose)
+	for i, job := range jobs {
+		if (i + 1) <= par {
+			spawn <- job
+		} else {
+			jobQueue <- job
+		}
 	}
 
-	res := []string{}
-	//wait loop
-	<-start
-	for e := range end {
-		more := false
+	globalResult := []string{}
+	expectedResultCount := initialJobCount
+	maxId := 0
+	for expectedResultCount > 0 {
```

Comment:
Don't we need to add a `|| len(spawn) > 0` also?

If `expectedResultCount` is 1, par is 1, and we get a new job to the `jobQueue`, when we get out the `singleResult`, we decrease the `expectedResultCount` to 0 and add a new element to the `spawn` and the for will stop while we still have an unprocessed element in `spawn`.

Reply:
Thanks a lot for giving it your additional thoughts!

Based on this I moved to counting running jobs which should be more robust and is easier to understand.
Update coming 😄

Reply:
Done, what do you think?